vars/xsDeploy.groovy

```diff
@@ -0,0 +1,328 @@
+import com.sap.piper.JenkinsUtils
+
+import static com.sap.piper.Prerequisites.checkScript
+
+import com.sap.piper.BashUtils
+import com.sap.piper.ConfigurationHelper
+import com.sap.piper.GenerateDocumentation
+import com.sap.piper.Utils
+
+import groovy.transform.Field
+
+import hudson.AbortException
+
+@Field String STEP_NAME = getClass().getName()
+
+@Field Set GENERAL_CONFIG_KEYS = STEP_CONFIG_KEYS
+
+@Field Set STEP_CONFIG_KEYS = [
+    'action',
+    'apiUrl',
+    'credentialsId',
+    'deploymentId',
+    'deployIdLogPattern',
+    'deployOpts',
+    /** A map containing properties forwarded to dockerExecute. For more details see [here][dockerExecute] */
+    'docker',
+    'loginOpts',
+    'mode',
+    'mtaPath',
+    'org',
+    'space',
+    'xsSessionFile',
+]
+
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+enum DeployMode {
+    DEPLOY,
+    BG_DEPLOY,
+    NONE
+
+    String toString() {
+        name().toLowerCase(Locale.ENGLISH).replaceAll('_', '-')
+    }
+}
+
+enum Action {
+    RESUME,
+    ABORT,
+    RETRY,
+    NONE
+
+    String toString() {
+        name().toLowerCase(Locale.ENGLISH)
+    }
+}
+
+/**
+  * Performs an XS deployment
+  */
+@GenerateDocumentation
+void call(Map parameters = [:]) {
+
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: parameters) {
+
+        def utils = parameters.juStabUtils ?: new Utils()
+
+        final script = checkScript(this, parameters) ?: this
+
+        ConfigurationHelper configHelper = ConfigurationHelper.newInstance(this)
+            .loadStepDefaults()
+            .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS)
+            .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+            .mixinStageConfig(script.commonPipelineEnvironment, parameters.stageName?:env.STAGE_NAME, STEP_CONFIG_KEYS)
+            .addIfEmpty('mtaPath', script.commonPipelineEnvironment.getMtarFilePath())
+            .addIfEmpty('deploymentId', script.commonPipelineEnvironment.xsDeploymentId)
+            .mixin(parameters, PARAMETER_KEYS)
+
+        Map config = configHelper.use()
+
+        DeployMode mode = config.mode
+
+        if(mode == DeployMode.NONE) {
+            echo "Deployment skipped intentionally. Deploy mode '${mode.toString()}'."
```

Comment:
[Q] Why is the `mode.toString()`? Wouln't it be sufficient to have simply `mode`?
[A] Unfortunately not. This works fine on the level of the unit test. But with the groovy environment available on Jenkins this does not work. On Jenkins we get crazy failures in this case. E.g. it is not clearly reported with an exception that there is something not working as expected. Instead we get some other class than a GSting which cannot be handled afterwards. Something similar we have with the stream APi which is also know for not working well on Jenkins ...